File: cs752/RTC/Code/Camera.cpp
/*
Author: Daniel Knutsen, Trevor Gerhardt
Date:   Spring 2010
Course: CS 752 
*/

#include "Camera.h"

#include "Point.h"
#include "Ray.h"
#include "Vector.h"

namespace RT {

   Point PinholeCamera::getEye() {
      return this->eye;
   }

   PinholeCamera::PinholeCamera( Point eye, Point lat, Vector up, float fov )
      : eye( eye ), lat( lat ), up( up ), fov( fov ) { }

   void PinholeCamera::preprocess( float aspect ) {

      this->L = (this->lat - this->eye).normal();
      Vector Ut = cross( this->L, this->up );
      Vector Vt = cross( Ut, this->L );

      //Vector U = (tan(deg2rad*fov*0.5f)) * Ut.normal();
      //Vector V = ((1/aspect)*tan(deg2rad*fov*0.5f)) * Vt.normal();
      this->U = Ut.normal();
      this->U = this->U * tan( deg2rad * this->fov * 0.5f );
      this->V = Vt.normal();
      this->V = this->V * ( ( 1/aspect ) * tan( deg2rad * this->fov * 0.5f ) );

   }

   Ray PinholeCamera::generateRay( float x, float y ) const {
      return Ray( this->eye, (this->L + (x * this->U) + (y * this->V) ).normal() );
   }

} // namespace RT

File: cs752/RTC/PcClient/PcClient/Core/Camera.cs
/*
Author: Daniel Knutsen, Trevor Gerhardt
Date:   Spring 2010
Course: CS 752 
*/

namespace RT {

    public class Camera {

        public double deg2rad = 0.0174532925f;
        public Point eye;
        public Point lat;
        public Vector up;
        public float fov;

        public Vector U, V, L;

        public Point getEye()
        {
            return this.eye;
        }

        public void preprocess(float aspect)
        {

            this.L = (this.lat - this.eye).normal();
            Vector Ut = Vector.cross( this.L, this.up );
            Vector Vt = Vector.cross( Ut, this.L );

            this.U = Ut.normal();
            this.U = this.U * (float)System.Math.Tan( this.deg2rad * this.fov * 0.5f );
            this.V = Vt.normal();
            this.V = this.V * ( ( 1/aspect ) * (float)System.Math.Tan( this.deg2rad * this.fov * 0.5f ) );

        }

        public Ray generateRay(float x, float y)
        {
            return new Ray( this.eye, (this.L + (x * this.U) + (y * this.V) ).normal() );
        }

    }

    class PinholeCamera : Camera
    {
        public PinholeCamera(Point eye, Point lat, Vector up, float fov)
        {
            this.eye = eye;
            this.lat = lat;
            this.up = up;
            this.fov = fov;  
        }
    }

} // namespace RT